processing/src/main/java/org/apache/druid/segment/UnnestStorageAdapter.java

```diff
@@ -226,9 +247,146 @@ public Metadata getMetadata()
     return baseAdapter.getMetadata();
   }
 
-  public String getDimensionToUnnest()
+  public VirtualColumn getUnnestColumn()
   {
-    return dimensionToUnnest;
+    return unnestColumn;
   }
-}
 
+  /**
+   * Split queryFilter into pre- and post-correlate filters.
+   *
+   * @param queryFilter            query filter passed to makeCursors
+   * @param queryVirtualColumns    query virtual columns passed to makeCursors
+   * @param inputColumn            input column to unnest if it's a direct access; otherwise null
+   * @param inputColumnCapabilites input column capabilities if known; otherwise null
+   *
+   * @return pair of pre- and post-correlate filters
+   */
+  private Pair<Filter, Filter> computeBaseAndPostCorrelateFilters(
+      @Nullable final Filter queryFilter,
+      final VirtualColumns queryVirtualColumns,
+      @Nullable final String inputColumn,
+      @Nullable final ColumnCapabilities inputColumnCapabilites
+  )
+  {
+    class FilterSplitter
+    {
+      final List<Filter> preFilters = new ArrayList<>();
+      final List<Filter> postFilters = new ArrayList<>();
+
+      void add(@Nullable final Filter filter)
+      {
+        if (filter == null) {
+          return;
+        }
+
+        final Set<String> requiredColumns = filter.getRequiredColumns();
+
+        // Run filter post-correlate if it refers to any virtual columns.
+        if (queryVirtualColumns.getVirtualColumns().length > 0) {
+          for (String column : requiredColumns) {
+            if (queryVirtualColumns.exists(column)) {
+              postFilters.add(filter);
+              return;
+            }
+          }
+        }
+
+        if (requiredColumns.contains(outputColumnName)) {
+          // Try to move filter pre-correlate if possible.
+          final Filter newFilter = rewriteFilterOnUnnestColumnIfPossible(filter, inputColumn, inputColumnCapabilites);
+          if (newFilter != null) {
+            preFilters.add(newFilter);
```

Comment:
actually I think we want to always add the filter as a post filter even when we push down since the pushed down filter returns the whole row if any values match, so we want the value matcher to clean it up to exact matches

Reply:
Let's merge this once CI passes, I'll fix them up. I understand the case where it returns the full row even for a match in MVDs. I'll also need to remove the allowSet and minor other changes. Will followup as soon as this gets merged

Reply:
Here's the patch https://gist.github.com/gianm/86d55fb6e5ee58c935152761a9df4a4d. Feel free to roll this into what you're doing @somu-imply.

Reply:
PR with the fix is at: https://github.com/apache/druid/pull/13919. If you want to incorporate this into your own work then feel free and I will close my PR.